File: src/middlewares/retry.ts
import type {
	DescribableInput,
	DescribableOutput,
	LanguageModelMiddleware,
	LanguageModelMiddlewareContext,
	LanguageModelMiddlewareNext,
} from "../describe.ts"

export const retry =
	<
		Input extends DescribableInput = DescribableInput,
		Output extends DescribableOutput = DescribableOutput,
	>(
		max: number,
	): LanguageModelMiddleware<Input, Output> =>
	async (
		context: LanguageModelMiddlewareContext<Input, Output>,
		next: LanguageModelMiddlewareNext<Input, Output>,
	) => {
		let last: unknown
		for (let i = 0; i < max; i++) {
			try {
				return await next(context)
			} catch (error) {
				last = error
			}
		}
		throw last
	}

File: go/middlewares/retry.go
package middlewares

import (
	"github.com/nmnmcc/toolbox-go"
)

// Retry creates a middleware that retries failed calls up to maxAttempts times.
func Retry[Input, Output any](
	maxAttempts int,
) toolbox.LanguageModelMiddleware[Input, Output] {
	return func(
		ctx toolbox.LanguageModelMiddlewareContext[Input, Output],
		next toolbox.MiddlewareNext[toolbox.LanguageModelMiddlewareContext[Input, Output], toolbox.LanguageModelCompletionContext[Input, Output]],
	) (toolbox.LanguageModelCompletionContext[Input, Output], error) {
		var lastErr error
		for i := 0; i < maxAttempts; i++ {
			result, err := next(ctx)
			if err == nil {
				return result, nil
			}
			lastErr = err
		}
		return toolbox.LanguageModelCompletionContext[Input, Output]{}, lastErr
	}
}
